lib/charms/sdcore_nrf/v0/fiveg_nrf.py

```diff
@@ -0,0 +1,281 @@
+        """Init."""
+        super().__init__(handle)
+        self.url = url
+
+    def snapshot(self) -> dict:
+        """Returns snapshot."""
+        return {"url": self.url}
+
+    def restore(self, snapshot: dict) -> None:
+        """Restores snapshot."""
+        self.url = snapshot["url"]
+
+
+class NRFRequirerCharmEvents(CharmEvents):
+    """List of events that the NRF requirer charm can leverage."""
+
+    nrf_available = EventSource(NRFAvailableEvent)
+
+
+class NRFRequires(Object):
+    """Class to be instantiated by the NRF requirer charm."""
+
+    on = NRFRequirerCharmEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    @staticmethod
+    def _relation_data_is_valid(relation_data: dict) -> bool:
+        """Returns whether URL is valid.
+
+        Args:
+            str: URL to be validated.
+
+        Returns:
+            bool: True if URL is valid, False otherwise.
+        """
+        try:
+            ProviderSchema(app=relation_data)
+            return True
+        except ValidationError:
+            return False
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handler triggered on relation changed event.
+
+        Args:
+            event: Juju event (RelationChangedEvent)
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation(self.relation_name)
+        if not relation:
+            logger.warning(f"No relation: {self.relation_name}")
+            return
+        relation = event.relation
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relation_name)
+            return
+        remote_app_relation_data = dict(relation.data[relation.app])
+        if not self._relation_data_is_valid(remote_app_relation_data):
+            logger.error("Invalid relation data: %s", remote_app_relation_data)
+            return
+        self.on.nrf_available.emit(url=remote_app_relation_data["url"])
+
+    def get_nrf_url(self) -> Optional[str]:
+        """Returns NRF url.
+
+        Returns:
+            str: NRF url.
+        """
+        relation = self.model.get_relation(self.relation_name)
+        if not relation:
+            logger.warning(f"No relation: {self.relation_name}")
+            return None
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relation_name)
+            return None
+        remote_app_relation_data = dict(relation.data[relation.app])
+        if not self._relation_data_is_valid(remote_app_relation_data):
+            logger.error("Invalid relation data: %s", remote_app_relation_data)
+            return None
```

Comment:
I think this part could go into a private function `_get_url` or _get_relation_data` as it is duplicated here and in the `_on_relation_changed`

Reply:
I agree, done.